### DISCUSSION TITLE
Setting required coupled variables through actions.

### DISCUSSION THREAD

**Discussion by leanderVS** [idaholab/moose#18110] [Q&A Modules: Phase field] [answered]
Hi everyone

An issue/question concerning the KKSMultiACBulkF kernel. 

Within this kernel, the variable that the kernel operates on has to be provided through the usual ‘variable = …’. The kernel also requires the user to specify a coupled variable ‘eta_i = …’ which defines the order parameter that derivatives are taken with respect to. However, within my (limited) experience, both of these variables are always the same, which makes the introduction of eta_i kind of redundant. Is there a reason the kernel does not simply use "variable" to define the derivatives? 

Of more importance to my own research is the actual definition of eta_i as a coupled variable. I modified the KKSMultiACBulkF kernel to generate a bulk free energy expression Sum_j[h(eta_j)F_j]. The F_j’s are constant (but different) at the moment, while the h(eta_j) were defined through SwitchingFunctionMultiPhaseMaterial. 

I am modelling an electrolyte in contact with a polycrystalline metal sample, which requires multiple of these adapted kernels to be called (each grain was assigned a different order parameter). When introduced manually into the inputfile, these give the expected results. However, altering the number of grains also requires quite some adaptations to the inputfile to introduce the correct bulk free energy expression as discussed above. To simply this, I tried to write an action. Whereas “variable” can easily be set through for example: 
params.set<NonlinearVariableName>("variable") = var_name;
The issue here is that I have not found a way to add/set the InputParameter object of the coupledVariable eta_j within that action since the coupledVar eta_j does not have the same definition within MOOSE as a simple NonlinearVariableName. CoupledVariableName and quite some other constructions I tried did not work unfortunately. 

I guess there are two options: 
Either I have to rewrite KKSMultiACBulkF to avoid the use of eta_j entirely or I need a way to set eta_j through an action. Preferably the latter if anyone has a solution. 

Hope I formulated this clearly, any help would be appreciated, thanks! 

Kind regards

Leander

**Comment by GiudGiud** [comment]
@dschwen @laagesen @amjokisaari if any of you could please answer this phase field question

**Comment by dschwen** [comment, marked as answer]
The variables are not always the same. This Kernel appears on the equation for the lagrange multiplier that enforces the sum of all switching functions to be 1 in the case of many (>2) phases.

**Reply by dschwen** [reply]
As for

> The issue here is that I have not found a way to add/set the InputParameter object of the coupledVariable eta_j within that action since the coupledVar eta_j does not have the same definition within MOOSE as a simple NonlinearVariableName. CoupledVariableName and quite some other constructions I tried did not work unfortunately

try `params.set<std::vector<VariableName>>("variable") = {var_name};`

A coupled variable can be an AuxVariable or a non linear Variable (hence the general `VariableName` rather than `NonlinearVariableName` (which you forgot in your example) or `AuxVariableName`), and MOOSE allows you to couple an arbitrary number of variables for each parameter (hence the `std::vector`).

**Reply by dschwen** [reply]
See here for an example of both uses:

https://github.com/idaholab/moose/blob/06bc1f73f0d4dfb3d01ce9ba77d6bb97346e07eb/modules/phase_field/src/action/ConservedAction.C#L173-L174

**Reply by leanderVS** [reply]
@dschwen , thank you for your reply!

I already had a look at ConservedAction.C but must have missed that particular part, thanks for pointing this out. The code now perfectly compiles and the simulation even converges, but it appears that it somehow still ignores the extra kernel I tried to add.

This is now the addition I made to PolycrystalKernelAction (now called CorrosionPolycrystalKernelAction for my purposes) to implement the bulk free energy terms (using the kernel CorrosionFb_I) (since I believe that in my case both variable and eta_i should refer to the same variable that the kernel operates on. 

    {
      InputParameters params = _factory.getValidParams("CorrosionFb_I");
      params.set<NonlinearVariableName>("variable") = var_name;
      params.set<std::vector<VariableName>>("args") = v;

      params.set<std::vector<VariableName>>("eta_i") = {var_name};

      params.set<std::vector<MaterialPropertyName>>("Fj_names") = _Fj_names;
      params.set<std::vector<MaterialPropertyName>>("hj_names") = _hj_names;
      params.applyParameters(parameters());

      std::string kernel_name = "CorrosionFb_I_" + var_name;
      _problem->addKernel("CorrosionFb_I", kernel_name, params);
    }

In addition, the typical 

params.addParam<std::vector<MaterialPropertyName>>("Fj_names","...")
params.addParam<std::vector<MaterialPropertyName>>("hj_names","...")

were added to the validParams() function of the action. Otherwise I think nothing changed for this action (apart from the declaration of Fj_names and hj_names in the .h file). 

The action is then called in the Inputfile using: 


[Kernels]
  [./CorrosionPolycrystalKernel]
// All different grain orientations have the same constant bulk free energy, the electrolyte (eta9) has a lower bulk free energy

    op_num = 10  
    var_name_base = eta 
    Fj_names = 'F_metal   F_metal   F_metal   F_metal   F_metal   F_metal   F_metal   F_metal   F_metal   F_elec'
    hj_names = 'h_metal_0 h_metal_1 h_metal_2 h_metal_3 h_metal_4 h_metal_5 h_metal_6 h_metal_7 h_metal_8 h_elec'

  [../]
[]

This implementation, however, appears to ignore the addition of the extra kernel. It thus gives the same results as if I wouldn't have implemented the extra kernel (although I did not overlap the two simulations to compare them one-to-one). 

As mentioned, the kernel does work on its own, when it is introduced as such: 

`
[Kernels]
  [./PolycrystalKernel]

    op_num = 10  
    var_name_base = eta  

  [../]

  //Kernels for bulk free energy introduction

  [./F_bulk_metal_0]

    type = CorrosionFb_I
    Fj_names = 'f_metal f_metal f_metal f_metal f_metal f_metal f_metal f_metal f_metal f_elec'
    eta_i = 'eta0'
    hj_names = 'h_metal_0 h_metal_1 h_metal_2 h_metal_3 h_metal_4 h_metal_5 h_metal_6 h_metal_7 h_metal_8 h_elec'
    variable = eta0
    args = 'eta1 eta2 eta3 eta4 eta5 eta6 eta7 eta8 eta9'

  []

  [./F_bulk_metal_1]
//Notice that all the metallic grain have the same constant free energy contribution. 

    type = CorrosionFb_I
    Fj_names = 'f_metal f_metal f_metal f_metal f_metal f_metal f_metal f_metal f_metal f_elec'
    eta_i = 'eta1'
    hj_names = 'h_metal_0 h_metal_1 h_metal_2 h_metal_3 h_metal_4 h_metal_5 h_metal_6 h_metal_7 h_metal_8 h_elec'
    variable = eta1
    args = 'eta0 eta2 eta3 eta4 eta5 eta6 eta7 eta8 eta9'

  []

  //and a lot more of the same stuff for all the other order parameters.

[]
`

I suppose the issue still lies in the action, but since the eta_i seems to be fine thanks to your fix, I ran out of ideas of where the issue could be. 

Best regards

Leander
